Fix test case backlog.population

clemahieu opened the pull request.

clemahieu committed:
Mutex must survive past nano::test::system.

dsiganos commented on nano/core_test/backlog.cpp:
```diff
@@ -72,10 +72,10 @@ block_list_t setup_independent_blocks (nano::test::system & system, nano::node &
  */
 TEST (backlog, population)
 {
```
Moving the mutex upwards does not fix the problem. The mutex is still destroyed before the node. The node is destroyed much after the test case comes to an end.

pwojcikdev replied:
I don't think this is true. `system` destructor does indeed stop the nodes, which means no node threads should be running after that

dsiganos commented:
Duplicate of #4059

clemahieu committed:
Stop must be called within ~backlog_population. If the system is destroyed and nodes have not been explicitly stopped, for instance in a test case failure, this would leave the backlog thread running.

clemahieu committed:
This fixes a race condition on stopping where the backlog_population thread might observe that the thread is not stopped but is not yet queued to receive thread notifications and might stall indefinitely.
Moving unlock during slow operation within populate_backlog since this loop also checks if it's been stoped.
Demoting stopped to a plain bool since it's used within locked contexts.
Moving activated container so it survives system destruction.

pwojcikdev commented on nano/node/backlog_population.cpp:
```diff
@@ -13,6 +13,7 @@ nano::backlog_population::backlog_population (const config & config_a, nano::sto
 
 nano::backlog_population::~backlog_population ()
 {
+	stop ();
```
This shouldn't be there, while it's not an error, we want to ensure that when node destruction happens all threads have already been stopped.

clemahieu committed:
Removing stop from destructor.
